File: 126_word_ladder_2.java
public class Solution {
    public List<List<String>> findLadders(String start, String end, Set<String> dict) {
        List<List<String>> ret = new ArrayList();
        if (dict == null || dict.isEmpty()) {
            return ret;
        }
        if (!dict.contains(start) || !dict.contains(end)) {
            return ret;
        }
        
        Map<String, List<String>> map = new HashMap();
        Map<String, Integer> distances = new HashMap();
        List<String> path = new ArrayList();
        bfs(map, distances, start, end, dict);
        dfs(ret, path, end, start, map, distances);
        return ret;
    }
    
    // bfs map for child has List<String> parent.
    public static void bfs(Map<String, List<String>> map, Map<String, Integer> distances,
                        String start, String end, Set<String> dict) {
        Queue<String> queue = new LinkedList();
        queue.add(start);
        distances.put(start, 0);
        for (String str : dict) {
            map.put(str, new ArrayList<String>());
        }
        
        while (!queue.isEmpty()) {
            String word = queue.poll();
            List<String> neighbors = expand(word, dict);
            for (String neighbor : neighbors) {
                map.get(neighbor).add(word);
                if (!distances.containsKey(neighbor)) {
                    distances.put(neighbor, distances.get(word)+1);
                    queue.add(neighbor);
                }
            }
        }
    }
    
    public static List<String> expand(String word, Set<String> dict) {
        List<String> neighbors = new ArrayList();
        for (int i = 0; i < word.length(); i++) {
            for (char c = 'a'; c <= 'z'; c++) {
                char[] ch = word.toCharArray();
                if (ch[i] == c) continue;
                ch[i] = c;
                String newWord = new String(ch);
                if (dict.contains(newWord)) {
                    neighbors.add(newWord);
                }
            }
        }
        return neighbors;
    }
    
    // trace from end -> start
    public static void dfs(List<List<String>> ret, List<String> path, String word,
                        String start, Map<String, List<String>> map, Map<String, Integer> distances) {
        if (word.equals(start)) {
            path.add(0, word);
            ret.add(new ArrayList<String>(path));
            path.remove(0);
            return;
        }
        
        for (String neighbor : map.get(word)) {
            if (distances.containsKey(neighbor) &&
                distances.get(word) == distances.get(neighbor) + 1) { // backward trace.
                path.add(0, word);  // add current word
                dfs(ret, path, neighbor, start, map, distances);  // dfs next round
                path.remove(0);
            }
        }
    }
}

File: 126_word_ladder_2.py
"""
 Given two words (start and end), and a dictionary, find all shortest transformation sequence(s) from start to end, such that:

    Only one letter can be changed at a time
    Each intermediate word must exist in the dictionary

For example,

Given:
start = "hit"
end = "cog"
dict = ["hot","dot","dog","lot","log"]

Return

  [
    ["hit","hot","dot","dog","cog"],
    ["hit","hot","lot","log","cog"]
  ]

Note:

    All words have the same length.
    All words contain only lowercase alphabetic characters.
"""
class Solution(object):
    def findLadders(self, start, end, dict):
        """
        :type start: str
        :type end: str
        :type dict: Set[str]
        :rtype: List[List[int]]

        idea:
            1. BFS: (stop @ end, shortest path)
                - build word distance from start (keep out the loop)
                - build word's neighbor words list
            2. DFS:
                - drill down from start. go through neighbor list
                - save the path and return
        """
        if dict == None or len(dict) == 0:
            return []

        if not start in dict or not end in dict:
            return []

        words_map, dist_map = self.bfs(start, end, dict)
        ret = []
        self.dfs(ret, [], start, end, words_map, dist_map)
        return ret

    def bfs(self, start, end, dict):
        done = False
        distance = 0
        words_map = {}
        dist_map = {}
        queue = [start]
        while len(queue) > 0 and not done:
            size = len(queue)
            for i in range(size):
                word = queue.pop(0)
                if word == end:
                    done = True

                neighbors = self.expand(word, dict)
                words_map[word] = neighbors
                dist_map[word] = distance
                queue += neighbors

            distance += 1

        return words_map, dist_map

    def expand(self, word, dict):
        neighbors = []
        chars = 'abcdefghijklmnopqrstwvxyz'
        for index in range(len(word)):
            for c in chars:
                if word[index] != c:
                    new_word = word[:index] + c + word[index+1:]
                    if new_word in dict:
                        neighbors.append(new_word)

        return neighbors

    def dfs(self, ret, path, word, end, words_map, dist_map):
        if word == end:
            path.append(word)
            ret.append(list(path))
            path.pop()
            return

        word_distance = dist_map[word]
        for neighbor in words_map[word]:
            neighbor_distance = dist_map[neighbor]
            if word_distance + 1 == neighbor_distance:
                path.append(word)
                self.dfs(ret, path, neighbor, end, words_map, dist_map)
                path.pop()



if __name__ == "__main__":
    start = "a"
    end = "c"
    dict = set(["a", "b", "c"])

    sol = Solution()
    print sol.findLadders(start, end, dict)

